File: Estacionamento.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define HORA 3600
#define DIA (24 * HORA)
#define MES (30 * DIA)
#define QUINZE_MIN 900

typedef enum { PEQUENO, MEDIO, GRANDE, CAMINHONETE, PASSEIO, CARGA } TipoVeiculo;

typedef struct {
    TipoVeiculo tipo;
    long inicio;
    long fim;
} Veiculo;

double calcularValor(Veiculo v, int juiz_de_fora) {
    long atual = (v.fim == 0) ? time(NULL) : v.fim;
    long periodo = atual - v.inicio;
    
    if (juiz_de_fora) {
        if (periodo <= HORA) {
            return 5.0 * ceil((double)periodo / QUINZE_MIN);
        }
    }
    
    switch (v.tipo) {
        case PASSEIO:
            if (periodo < 12 * HORA) {
                return 2.0 * ceil((double)periodo / HORA);
            } else if (periodo >= 12 * HORA && periodo < 15 * DIA) {
                return 26.0 * ceil((double)periodo / DIA);
            } else {
                return 300.0 * ceil((double)periodo / MES);
            }
        case CARGA:
            return 50.0 * ceil((double)periodo / DIA);
        case PEQUENO:
            return 1.5 * ceil((double)periodo / HORA);
        case MEDIO:
            return 2.5 * ceil((double)periodo / HORA);
        case GRANDE:
            return 3.5 * ceil((double)periodo / HORA);
        case CAMINHONETE:
            return 4.5 * ceil((double)periodo / HORA);
        default:
            return 0;
    }
}

int main() {
    Veiculo v;
    int tipo, juiz_de_fora;
    printf("Digite o tipo de veículo (0 = Pequeno, 1 = Médio, 2 = Grande, 3 = Caminhonete, 4 = Passeio, 5 = Carga): ");
    scanf("%d", &tipo);
    v.tipo = (TipoVeiculo)tipo;
    
    printf("Digite o horário de entrada (timestamp em segundos): ");
    scanf("%ld", &v.inicio);
    
    printf("Digite o horário de saída (timestamp em segundos, 0 para agora): ");
    scanf("%ld", &v.fim);
    
    printf("Aplicar regra de Juiz de Fora? (1 = Sim, 0 = Não): ");
    scanf("%d", &juiz_de_fora);
    
    double valor = calcularValor(v, juiz_de_fora);
    printf("Valor a pagar: R$ %.2f\n", valor);
    
    return 0;
}

File: Estacionamento.py
from datetime import datetime, timedelta
import math

HORA = 3600
DIA = HORA * 24
MES = DIA * 30
QUINZE_MIN = 900

class TipoVeiculo:
    PEQ = 1
    MED = 2
    GRA = 3
    CAM = 4

class ContaEstacionamento:
    def __init__(self, tipo, hora_entrada, hora_saida):
        self.tipo = tipo
        self.hora_entrada = hora_entrada
        self.hora_saida = hora_saida

def regra_juiz_de_fora(veiculo):
    permanencia = veiculo.hora_saida - veiculo.hora_entrada
    permanencia_seg = permanencia.total_seconds()

    if permanencia_seg < HORA:
        return 0.5 * math.ceil(permanencia_seg / QUINZE_MIN)
    else:
        return calcDiaria(veiculo)

def calcDiaria(veiculo):
    permanencia = veiculo.hora_saida - veiculo.hora_entrada
    perm = permanencia.total_seconds()

    if veiculo.tipo == TipoVeiculo.PEQ:
        if perm <= 12 * HORA:
            return 2.0 * math.ceil(perm / HORA)
        elif perm <= 15 * DIA:
            return 26.0 * math.ceil(perm / DIA)
        else:
            return 300.0 * math.ceil(perm / MES)
            
    elif veiculo.tipo == TipoVeiculo.MED:
        return 2.5 * math.ceil(perm / HORA)
    elif veiculo.tipo == TipoVeiculo.GRA:
        return 3.5 * math.ceil(perm / HORA)
    elif veiculo.tipo == TipoVeiculo.CAM:
        return 4.5 * math.ceil(perm / HORA)
    
    return 0.0  # Caso o tipo seja inválido

def main():
    tipo = int(input("Digite o tipo de veículo (1 = Pequeno, 2 = Médio, 3 = Grande, 4 = Caminhonete): "))
    
    hora_entrada = input("Insira a hora de entrada (HH:MM): ").strip()
    hora_saida = input("Insira a hora de saída (HH:MM): ").strip()

    formato = "%H:%M"
    entrada = datetime.strptime(hora_entrada, formato)
    saida = datetime.strptime(hora_saida, formato)

    if saida <= entrada:
        print("Erro: A hora de saída deve ser posterior à hora de entrada.")
        return

    aplicar_regra_jf = int(input("Aplicar regra de Juiz de Fora? (1 = Sim, 0 = Não): "))

    veiculo = ContaEstacionamento(tipo, entrada, saida)

    if aplicar_regra_jf == 1:
        valor = regra_juiz_de_fora(veiculo)
    else:
        valor = calcDiaria(veiculo)

    print(f"Valor a pagar: R$ {valor:.2f}")

if __name__ == "__main__":
    main()